File: 21-MergeTwoSortedLists/21-MergeTwoSortedLists.py
# Last updated: 6/25/2026, 9:17:59 AM
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution(object):
    def mergeTwoLists(self, list1, list2):
        """
        :type list1: Optional[ListNode]
        :type list2: Optional[ListNode]
        :rtype: Optional[ListNode]
        """
        head1=list1
        head2=list2
        res=ListNode()
        curr=res
        while head1 and head2:
            if head1.val>= head2.val:
                curr.next=head2
                head2=head2.next
            else:
                curr.next=head1
                head1=head1.next
            curr=curr.next
        
        while head1:
            curr.next=head1
            curr=curr.next
            head1=head1.next
        
        while head2:
            curr.next=head2
            curr=curr.next
            head2=head2.next
        
        return res.next
            

File: 21-MergeTwoSortedLists/21-MergeTwoSortedLists.cpp
// Last updated: 6/25/2026, 9:18:04 AM
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeTwoLists(ListNode* list1, ListNode* list2) {
        if(list1 == NULL) return list2;
        if(list2 == NULL) return list1;
        ListNode* trav; ListNode* trav2;
        if(list1->val > list2->val) {trav = list2; list2=list2->next;}
        else {trav = list1; list1=list1->next;}
        trav2 = trav;
        while(list2 != NULL and list1 != NULL){
            if(list1->val > list2->val) {trav->next = list2; trav=trav->next; list2=list2->next;}
            else {trav->next = list1; trav=trav->next; list1=list1->next;}
        }
        if(list2==NULL) trav->next = list1;
        else trav->next = list2;
        return trav2;
    }
};